rclpy/src/rclpy/utils.cpp

```diff
@@ -89,4 +97,103 @@ convert_to_py(void * message, py::object pyclass)
   return py::reinterpret_steal<py::object>(convert(message));
 }
 
+const char *
+get_rmw_implementation_identifier()
+{
+  return rmw_get_implementation_identifier();
+}
+
+void
+assert_liveliness(py::object pyentity)
+{
+  if (PyCapsule_IsValid(pyentity.ptr(), "rclpy_publisher_t")) {
+    auto publisher = static_cast<rclpy_publisher_t *>(
+      rclpy_handle_get_pointer_from_capsule(
+        pyentity.ptr(), "rclpy_publisher_t"));
```

Comment:
`rclpy_handle_get_pointer_from_capsule` may return a `nullptr`. If so it means the function has thrown a Python exception and this needs to do `throw py::error_already_set()` to tell pybind11 that happened.